[tt-train] SwiGLU forward kernel implementation

mdragulaTT opened the pull request:
## Ticket
[[Feature Request] SwiGLU forward and backward functions](https://github.com/tenstorrent/tt-metal/issues/14195)

## Problem description
This PR introduces a **dual-path SwiGLU forward operation** implementation with conditional compilation based on input matrix dimensions and L1 memory capacity. Currently, models depend on non-fused implementations of SwiGLU, which can be suboptimal for performance. This implementation provides both a memory-optimized fast path and a fallback streaming approach, establishing the foundation for performant fused SwiGLU while maintaining compatibility across different input sizes.

## What's changed
- **Added L1-optimized SwiGLU forward with conditional compilation**
  - **Fast path**: Caches full rows of XW1, XW3, and M in L1 when input dimensions allow the required memory to fit within fixed L1 capacity, eliminating redundant DRAM reads and recomputation
  - **Fallback path**: Streaming approach for larger input dimensions where full rows don't fit in L1, recomputes M as needed
  - Runtime memory checking automatically selects optimal algorithm based on whether memory required by input dimensions fits within fixed L1 capacity

- **Implementation details**
  - Uses block-based compute kernel
  - Enhanced tile loading utilities with row-major and column-major read functions
  - Conditional compilation with define flag propagation
  - Assumes both `C` (input dimension) and `hidden_C` (hidden dimension) are divisible by 32
  - Masking support for non-divisible dimensions **not yet implemented**

- **Backward path**
  - Added composite implementation for SwiGLU backward for compatibility with other ops

- **Testing**
  - Added comprehensive test suite using xtensor-based reference implementation to validate correctness across various scenarios

- **Limitations and future work**
  - This version does not utilize unicast/multicast optimizations (planned for next PR)
  - Implement masking for dimensions where `C % 32 != 0` or `hidden_C % 32 != 0`

## Performance
Current benchmarking on nanoGPT (7 steps, 36 calls, skipping first step):

**Composite implementation:**
- Mean: 6.808 ms
- Std dev: 0.028 ms

**Fused implementation (fast path):**
- Mean: 6.827 ms  
- Std dev: 0.069 ms

The fused implementation currently performs roughly equivalent to the composite version. Performance optimizations including unicast/multicast will be addressed in subsequent PRs.

**Methodology:** Performance measured using profiler markers wrapped around SwiGLU operations in `llama_block.cpp`, forward pass only with gradients disabled.

## Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

mdragulaTT committed:
Implement SwiGLU forward operation with block-based approach

- Added initial implementation of the SwiGLU forward op using a block-based compute strategy.
- This version assumes both input (C) and hidden (hidden_C) dimensions are divisible by 32; masking for non-divisible cases is not yet supported.
- Each value of the intermediate M is recomputed as needed, without L1 caching or reuse.
- Includes a composite implementation of the SwiGLU backward op for compatibility with other ops.
- Added a comprehensive test suite using an xtensor-based reference implementation for correctness.
- Future work: add masking for non-divisible dimensions and optimize memory access by caching rows of M or X in L1.

rfurko-tt commented on tt-train/sources/ttml/ops/swiglu_op.cpp:
```diff
@@ -0,0 +1,129 @@
+                    ttnn::subtract(
+                        ttnn::ones_like(sigmoid_linear1),
+                        sigmoid_linear1,
+                        std::nullopt,
```
are you sure you need to pass all this parameters?

rfurko-tt commented on tt-train/sources/ttml/metal/ops/swiglu_fw/device/swiglu_fw_program_factory.cpp:
```diff
@@ -0,0 +1,376 @@
+    uint32_t num_cores_y = compute_with_storage_grid_size.y;
+
+    // Compile arguments
+    // We enforce to use block_size of 4. If num_inner % 4 != 0 or hidden_num_inner % 4 != 0, we will take care of it in
```
👍

rfurko-tt commented on tt-train/sources/ttml/metal/ops/swiglu_fw/device/swiglu_fw_program_factory.cpp:
```diff
@@ -0,0 +1,376 @@
+    // - input CBs must be bf16
+    // - output CBs must be bf16
+    // - matmul_tiles supports only one data format for both input and output CBs (otherwise produces some NaNs/infs)
+    // - matmul_tiles and accumulation is done on FPU, which supports only bfloat16
```
based on discussion with Borys this statement is not true

rfurko-tt commented on tt-train/sources/ttml/metal/ops/swiglu_fw/swiglu_fw.cpp:
```diff
@@ -0,0 +1,25 @@
+    const ttnn::Tensor& input_tensor, const ttnn::Tensor& w1, const ttnn::Tensor& w2, const ttnn::Tensor& w3) {
+    return ttnn::prim::ttml_swiglu_fw(
+        input_tensor,  // [B,1,S,C]
+        w1,            // [C,H]
```
isn't it also 4D? [1, 1, C, H]

rfurko-tt commented on tt-train/sources/ttml/metal/ops/swiglu_fw/device/swiglu_fw_device_operation.cpp:
```diff
@@ -0,0 +1,129 @@
+    const auto& input_logical_shape = input_tensor.logical_shape();
+    auto program_factory = select_program_factory(args, tensor_args);
+    tt::tt_metal::operation::Hash hash = tt::tt_metal::operation::hash_operation<SwiGLUForwardDeviceOperation>(
+        args, program_factory.index(), input_tensor.dtype(), input_logical_shape);
```
should we also include shape of w1, w2 or w3? 

it's not a problem for now, but might happen if we have multiple models with different intermediate dimension size but same input size

rfurko-tt commented on tt-train/sources/ttml/metal/ops/swiglu_fw/device/kernels/dataflow/reader_swiglu_fw_interleaved_start_id.cpp:
```diff
@@ -0,0 +1,192 @@
+    const DataFormat data_fmt = get_dataformat(cb_input_idx);
+
+    // Address generators
+    const InterleavedAddrGenFast<true> x_ag = {
```
Use Stas instructions to refactor this code

rfurko-tt commented on tt-train/sources/ttml/metal/ops/swiglu_fw/device/kernels/dataflow/reader_swiglu_fw_interleaved_start_id.cpp:
```diff
@@ -0,0 +1,192 @@
+constexpr bool do_mask_hw = false;
+#endif
+
+// TODO(maciek): Decide on one consistent version of whitespace in X[a, b] vs. X[a,b]
```
i would prefer X[a, b]

rfurko-tt commented on tt-train/sources/ttml/metal/ops/swiglu_fw/device/kernels/compute/swiglu_fw_kernel.cpp:
```diff
@@ -0,0 +1,223 @@
+                        // Push empty/invalid M for k >= k_block_size
+                        tile_regs_acquire();
+                        tile_regs_commit();
+                        pack_and_push(3U, cb_m_idx);  // push empty/invalid M
```
i think this can be simplified. wdyt? 

```

for (uint32_t k = 0; k < k_block_size; ++k) {
   compute_M_for_k();
}
if (k_block_size != block_size) {
   // question, do not insert it in the code: 
   // you do syncronization to get zeros in registers? 
   tile_regs_acquire();
   tile_regs_commit();
   pack_and_push_block(cb_m_idx, block_size - k_block_size); 
}

```

rfurko-tt commented on tt-train/sources/ttml/metal/ops/swiglu_fw/device/kernels/compute/swiglu_fw_kernel.cpp:
```diff
@@ -0,0 +1,223 @@
+    tile_regs_commit();
+    // Store result to appropriate CB
+    const auto output_cb_idx = last_k_block ? cb_y_idx : cb_y_partial_idx;
+    pack_and_push_block(output_cb_idx, block_size);
```
the problem with this functions `pack_and_push*` that you always call reconfig data format, i believe that you don't need that. Can we add additional parameter to the function (it can be true by default) which is true if we need to reconfigure data format and false otherwise. 

would be interesting to see performance comparison

rfurko-tt reviewed (commented):
Overall looks very readable and cool. Great job @mdragulaTT ! 

Lets finish this kernel and move to backward :) Extremely excited

mdragulaTT committed:
Addressed comments

- Changed `w` tensors to be stored as 4D instead of 2D for consistency with model format
- Updated and clarified related code comments
- Included `w` tensor shapes in kernel hash computation for more robust cache keying
- Updated address generators to use `TensorAccessor`

mdragulaTT committed:
Added L1-optimized SwiGLU with conditional compilation

- Dual-path algorithm: L1-cached vs streaming based on memory availability.
- Runtime memory checking automatically selects optimal path.
- Fast path eliminates redundant DRAM reads by caching full M rows.
- Improved CB indexing and tile loading utilities.

mdragulaTT committed:
Flash-attention optimization for SwiGLU XW1/XW3 computation

- Reorder reader loops to process W1 data first, then W3 data per k_block
- Implement flash-attention pattern: read X[r, p_block] once and reuse
  across k_blocks by avoiding redundant X data fetching

rfurko-tt commented on tt-train/sources/ttml/metal/ops/swiglu_fw/device/swiglu_fw_program_factory.cpp:
```diff
@@ -0,0 +1,417 @@
+    // Compile arguments
+    // We enforce to use block_size of 4. If num_inner % 4 != 0 or hidden_num_inner % 4 != 0, we will take care of it in
+    // the kernels.
+    const uint32_t block_size = 4U;
```
can we do block_size = min(4U, Wt) ? wdyt? 
i can be wrong here

rfurko-tt reviewed (approved):
Overall looks great!

1. Update PR description 
2. remove compose operation option in new swiglu
3. Think about splitting compute into two separate kernels and choose which to use in program factory

mdragulaTT commented on tt-train/tests/ops/swiglu_op_test.cpp:
```diff
@@ -0,0 +1,192 @@
+}
+
+// 7. NanoGPT-like shape: 32x1x256x384, hidden_dim=1024
+TEST_F(SwiGLUOpTest, SwiGLU_NanoGPT_32x1x256x384) {
```
Consider `NIGHTLY_*`

zbaczewskiTT commented on tt-train/sources/ttml/metal/ops/swiglu_fw/device/swiglu_fw_program_factory.cpp:
```diff
@@ -0,0 +1,417 @@
+    // Compile arguments
+    // We enforce to use block_size of 4. If num_inner % 4 != 0 or hidden_num_inner % 4 != 0, we will take care of it in
+    // the kernels.
+    const uint32_t block_size = 4U;
```
Did you mean `/*fp32_dest_acc_en=*/false`, since it's already set to true?
We cannot set block size 8U with fp32_dest_acc_en=true (unless we disable double-buffering).
My understanding is that code not allowing to set fp32_desc_acc_en=false and block_size to 8U is this:
[L241-L269](https://github.com/tenstorrent/tt-metal/blob/5ae1d08109fe59a3ad3ced536c05445eaa2fd268/tt-train/sources/ttml/metal/ops/swiglu_fw/device/kernels/compute/swiglu_fw_kernel.cpp#L241-#L269)

The snippet is the only one executing consecutive SFPU operation - thus the only one benefitting from fp32 registers. We could use intermediate CB's (which you've mentioned in your note) and process the three operations (sigmoid + 2 elwmuls) in full blocks, by using FPU for elwmuls. With this change we could set fp32_desc_acc_en=false and block_size=8U, further increasing throughput. Of course using FPU for elwmuls could result in precision loss (Also sigmoid packed into bf16 register), but I imagine the precision problems in this kernel stem from matmuls rather than these operations.

I imagine the main performance issue in the kernel is not this thought - still maybe worth exploring since it increases overall throughput.

mdragulaTT commented on tt-train/sources/ttml/ops/swiglu_op.cpp:
```diff
@@ -0,0 +1,103 @@
+
+    ttnn::Tensor swiglu_fw_result;
+
+    if (use_composite_fw) {
```
Done.

mdragulaTT commented on tt-train/sources/ttml/metal/ops/swiglu_fw/device/swiglu_fw_program_factory.cpp:
```diff
@@ -0,0 +1,415 @@
+    // NOTE(maciek):
+    // - fp32 input/output CBs are possible, but here both are always bf16 to match pipeline formats.
+    // - matmul_tiles seems to require matching input/output CB dtypes, otherwise NaNs/infs may occur.
+    //   TODO(maciek): make minimal repro and report if this is a bug.
```
Done.

mdragulaTT committed:
Address code review comments

- Split compute kernels into separate 2 kernels
- Remove composite operation from SwiGLU forward op
- Fix typos in comments
- Link TODO comments with corresponding GitHub issues
- Add 2 additional test cases for better coverage

mdragulaTT committed:
Replace SwiGLU test tolerance with L2 norm and use parallel_generate for random data

- Switch from elementwise allclose() to relative L2 error validation for better scale-invariance
- Replace xt::random with core::parallel_generate following RMSNorm pattern
- Use autograd context RNG for consistent seeding across test runs
